lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -1344,6 +1337,38 @@ migrateWallet srcCtx sheCtx (ApiT srcWid) (ApiT sheWid) migrateData = do
             #pendingSince
   where
     pwd = coerce $ getApiT $ migrateData ^. #passphrase
+    addrs = getApiT . fst <$> migrateData ^. #addresses
+
+-- | Transform the given set of migration coin selections (for a source wallet)
+--   into a set of coin selections that will migrate funds to the specified
+--   target addresses.
+--
+-- Each change entry in the specified set of coin selections is replaced with a
+-- corresponding output entry in the returned set, where the output entry has a
+-- address from specified addresses.
+--
+-- The difference between the size of coin selection and the number of
+-- specified addresses is solved by cycling specified addresses.
+--
+changeCoinSelectionForMigration
+    :: [Address]
+    -- ^ Target addresses
+    -> [CoinSelection]
+    -- ^ Migration data for the source wallet.
+    -> [UnsignedTx]
+changeCoinSelectionForMigration addrs selections =
+    zipWith changeSel selections addrsPools
+  where
+      addrsPools =
+          fst $
+          foldl (\(res,addrPool) sel ->
+                     let len = length (change sel) in
+                         (take len addrPool:res, drop len addrPool)
+                )
+          ([], cycle addrs) selections
+      changeSel sel addrPool = UnsignedTx
+          (NE.fromList (sel ^. #inputs))
+          (NE.fromList (zipWith TxOut addrPool (sel  ^. #change)))
```

Comment:
**Suggestion**: simplify this function by iterating through the selection list just once:

```suggestion
    foldr accumulate ([], cycle addrs) selections
  where
    accumulate sel (txs, addrsAvailable) = first
        (\addrsSelected -> makeTx sel addrsSelected : txs)
        (splitAt (length $ change sel) addrsAvailable)
    makeTx :: CoinSelection -> [Address] -> UnsignedTx
    makeTx sel addrsSelected = UnsignedTx
        (NE.fromList (sel ^. #inputs))
        (NE.fromList (zipWith TxOut addrsSelected (sel ^. #change)))
```
(Note that `first` is taken from `Data.Bifunctor`.)

Also, using `foldr` here means that the order of selections is preserved, which might make this easier to test.

Reply:
indeed, it reads better and looks better ;-) changed (with small addition to make it work really) in https://github.com/input-output-hk/cardano-wallet/pull/1664/commits/cf2b9e58c71fd7e80f8a9e1b9073348dfba3a67e